service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/CaseExtensionController.java

```diff
@@ -0,0 +1,94 @@
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.service.ValidateGroupService;
+import uk.gov.hmcts.reform.fpl.validation.groups.CaseExtensionGroup;
+
+import java.time.LocalDate;
+
+import static org.springframework.util.StringUtils.isEmpty;
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.DATE;
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.formatLocalDateToString;
+
+
+@Api
+@RestController
+@RequestMapping("/callback/case-extension")
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class CaseExtensionController {
+    private final ObjectMapper mapper;
+    private final ValidateGroupService validateGroupService;
+    private LocalDate caseCompletionDate;
+    private LocalDate eightWeekExtensionDate;
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStartEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        if (isEmpty(caseData.getCaseCompletionDate())) {
+            caseCompletionDate = caseData.getDateSubmitted();
+        } else {
+            caseCompletionDate = caseData.getCaseCompletionDate();
+        }
+
+        caseDetails.getData().put("shouldBeCompletedByDate", formatLocalDateToString(caseCompletionDate, DATE));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackrequest) {
+        CaseDetails caseDetails = callbackrequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        eightWeekExtensionDate = caseCompletionDate.plusWeeks(8);
+
+        caseDetails.getData().put("extensionDateEightWeeks", formatLocalDateToString(eightWeekExtensionDate,
+                DATE));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .errors(validateGroupService.validateGroup(caseData, CaseExtensionGroup.class))
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+
+        caseCompletionDate = getCaseCompletionDate(caseDetails);
+        caseDetails.getData().put("caseCompletionDate", caseCompletionDate);
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    private LocalDate getCaseCompletionDate(CaseDetails caseDetails) {
```

Comment:
Might be good idea to create CaseExtensionService and put this method with other pieces of logic there. We could unit test it then, so we wouldn't need so many ITs.

Reply:
done :)